### PR TITLE
[GEOT-7288] upgrade javax.measure version to 2.1.3

### PR DESCRIPTION
**Pull request by AlbanSeurat**
[GEOT-7288](https://osgeo-org.atlassian.net/browse/GEOT-7288) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

Upgrade the javax.measure library
  

# Checklist

- [X] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [ ] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [X] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [X] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [x] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [X] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [ ] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [ ] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [X] Bug fixes and small new features are presented as a single commit.
- [X] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

**Comment by AlbanSeurat** [comment]
Not sure I need the to sign the contribution as this is really smart change. Let me know if you need it

**Comment by ianturton** [comment]
LGTM, assuming the tests pass

**Comment by AlbanSeurat** [comment]
Seems there is some formatting issue, I'll run mvn spotless:apply and republished the Pull Request

**Comment by AlbanSeurat** [comment]
I have an error here : 

java.lang.IllegalArgumentException: Layer group must not be empty

	at org.geoserver.catalog.impl.CatalogImpl.validate(CatalogImpl.java:1021)
	at org.geoserver.catalog.impl.CatalogImpl.add(CatalogImpl.java:954)
	at org.geoserver.security.SecureCatalogImpl.add(SecureCatalogImpl.java:1260)
	at org.geoserver.catalog.impl.AbstractFilteredCatalog.add(AbstractFilteredCatalog.java:573)
	at org.geoserver.catalog.impl.AbstractCatalogDecorator.add(AbstractCatalogDecorator.java:532)
	at org.geoserver.catalog.impl.LocalWorkspaceCatalog.add(LocalWorkspaceCatalog.java:412)
	at org.geoserver.catalog.CatalogIntegrationTest.testCascadeDeleteWorkspaceSpecific(CatalogIntegrationTest.java:316)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)

Any idea how my changes could impact that ?

**Comment by aaime** [comment]
I quickly glanced over and test expectations seems to have been changed.
That might be a problem, but need more time to check and verify.

Il Gio 1 Dic 2022, 02:16 Ian Turton ***@***.***> ha scritto:

**Comment by AlbanSeurat** [comment]
I figure out where was the issue was. systems-common 2.1 was using indriya 2.1.2 whereas I was using in code the indriya 2.1.3 version (SimplerUnitFormmater.getNewInstance()).

**Comment by AlbanSeurat** [comment]
Hello, I have addressed the remaining issue regarding the geoserver build. Any chances to have the pull request merged in the coming days ? Is there any tasks I can help on to move forward the PR ?

**Comment by aaime** [comment]
Not this week, I hope to find time to review it Monday 12th, but remember, review is a volunteer activity, there are no guarantees as to when it might happen.

File: modules/library/metadata/src/main/java/org/geotools/measure/BaseUnitFormatter.java
Comment (aaime) [inline modules/library/metadata/src/main/java/org/geotools/measure/BaseUnitFormatter.java:91]: Public API in a public class cannot be removed like this. It has to be kept functional, and deprecated, for a release cycle, and only then it can be removed.
The change should also be noted in the upgrade guide here: https://github.com/geotools/geotools/blob/main/docs/user/welcome/upgrade.rst

File: modules/library/metadata/src/test/java/org/geotools/measure/UnitFormatterTest.java
Comment (aaime) [inline modules/library/metadata/src/test/java/org/geotools/measure/UnitFormatterTest.java:187]: This is a problem in Java 17 (should cause a runtime warning). Please try to find a way that does not leverage reflection, as we'll eventually mandate Java 17 as the minim version.

File: modules/plugin/coverage-multidim/netcdf/src/test/java/org/geotools/coverage/io/netcdf/NetCDFUnitParserTest.java
Comment (aaime) [inline modules/plugin/coverage-multidim/netcdf/src/test/java/org/geotools/coverage/io/netcdf/NetCDFUnitParserTest.java:171]: Any chance you can build tolerance for "degree" in the library, to avoid backward compatibility issues?

**Review by aaime — CHANGES_REQUESTED** [review]
Generally looks good, but there are a couple of serious issues in the changes.

File: modules/plugin/coverage-multidim/netcdf/src/test/java/org/geotools/coverage/io/netcdf/NetCDFUnitParserTest.java
Comment (AlbanSeurat) [reply modules/plugin/coverage-multidim/netcdf/src/test/java/org/geotools/coverage/io/netcdf/NetCDFUnitParserTest.java:171]: My mistake , "degree" still works. I revert my change.

File: modules/library/metadata/src/test/java/org/geotools/measure/UnitFormatterTest.java
Comment (AlbanSeurat) [reply modules/library/metadata/src/test/java/org/geotools/measure/UnitFormatterTest.java:187]: removed

File: modules/library/metadata/src/main/java/org/geotools/measure/BaseUnitFormatter.java
Comment (AlbanSeurat) [reply modules/library/metadata/src/main/java/org/geotools/measure/BaseUnitFormatter.java:91]: I thought it was only used by tests and therefore not necessary. I put them back

**Comment by AlbanSeurat** [comment]
Should I rewrite history after the review or create a new commit ? I put a new commit but I am ready to squash before the merge.

**Comment by aaime** [comment]
We'll do a "squash-merge", no worries.

File: modules/library/metadata/src/main/java/org/geotools/measure/BaseUnitFormatter.java
Comment (ianturton) [inline modules/library/metadata/src/main/java/org/geotools/measure/BaseUnitFormatter.java:21]: This will (should) fail the PMC checks as we don't allow wildcard imports.

File: modules/library/metadata/src/main/java/org/geotools/measure/BaseUnitFormatter.java
Comment (ianturton) [inline modules/library/metadata/src/main/java/org/geotools/measure/BaseUnitFormatter.java:193]: Can you make these into constants too?

**Review by ianturton — CHANGES_REQUESTED** [review]

File: modules/library/metadata/src/main/java/org/geotools/measure/BaseUnitFormatter.java
Comment (AlbanSeurat) [reply modules/library/metadata/src/main/java/org/geotools/measure/BaseUnitFormatter.java:21]: I ran mvn spotless:apply. Does it not change that ?

Comment (AlbanSeurat) [reply modules/library/metadata/src/main/java/org/geotools/measure/BaseUnitFormatter.java]: Change my IDE setting

File: modules/library/metadata/src/main/java/org/geotools/measure/BaseUnitFormatter.java
Comment (AlbanSeurat) [reply modules/library/metadata/src/main/java/org/geotools/measure/BaseUnitFormatter.java]: Done

**Review by ianturton — APPROVED** [review]

**Comment by soc** [comment]
Saw this change due to internal test failures. There was a reason why this version was not used when the code was written:

`SimpleUnitFormat.getNewInstance()` does not get you a "fresh" instance, but an instance pre-initialized with whatever definitions that Indriya library came up with.

You now pretty much bet GeoTools' house on the state of that initialization, or the hope that the initialization done by GeoTools overwrites everything of relevance. (Not sure if the API does actually overwrite or just adds more units/labels.)

Sometimes it makes sense to ask "why?", before upping version numbers for the sake of upping version numbers.

**Comment by AlbanSeurat** [comment]
@soc How can I help on your tests as I am the one who changed the code ?

**Comment by soc** [comment]
It is what it is. ¯\\\_(ツ)\_/¯

When I did the original implementation, I knew that newer versions were unable to implement the semantics due to API changes in later minor versions. So I wrote explicit version check against the Unit library back then, and that version check triggered with the GT update, confirming my predictions.

**Pull request merged**
